src/anyio/_core/_tasks.py

```diff
@@ -156,3 +180,248 @@ def create_task_group() -> TaskGroup:
+
+
+class TaskHandle(Generic[T]):
+    __slots__ = (
+        "cancel_scope",
+        "_event",
+        "_return_value",
+        "_exception",
+        "_start_value",
+    )
+
+    _return_value: T
+    _start_value: Any
+
+    def __init__(self) -> None:
+        from ._synchronization import Event
+
+        self.cancel_scope: CancelScope = CancelScope()
+        self._event = Event()
+        self._exception: BaseException | None = None
+
+    def set_return_value(self, value: T) -> None:
+        self._return_value = value
+        self._event.set()
+
+    def set_exception(self, exception: BaseException) -> None:
+        self._exception = exception
+        self._event.set()
+
+    def cancel(self) -> None:
+        self.cancel_scope.cancel()
+
+    @property
+    def start_value(self) -> Any:
+        try:
+            return self._start_value
+        except AttributeError:
+            raise ValueError(
+                "this task handle has no start value (the task was not started with "
+                "'await tg.start(...)`)"
+            )
+
+    async def wait(self) -> T:
+        await self._event.wait()
+        if self._exception is not None:
+            raise self._exception
+
+        return self._return_value
+
+
+class EnhancedTaskGroup:
+    async def __aenter__(self) -> Self:
+        self._task_group = create_task_group()
+        await self._task_group.__aenter__()
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> None:
+        await self._task_group.__aexit__(exc_type, exc_val, exc_tb)
+
+    @staticmethod
+    async def _run_coro(coro: Coroutine[Any, Any, T], handle: TaskHandle[T]) -> None:
+        __tracebackhide__ = True
+        with handle.cancel_scope:
+            try:
+                retval = await coro
+            except BaseException as exc:
+                handle.set_exception(exc)
+            else:
+                handle.set_return_value(retval)
+
+    @staticmethod
+    async def _run_start_func(
+        func: Callable[..., Coroutine[Any, Any, T]],
+        args: tuple[Any],
+        kwargs: Mapping[str, Any],
+        handle: TaskHandle[T],
+        *,
+        task_status: TaskStatus[Any],
+    ) -> None:
+        __tracebackhide__ = True
+        with handle.cancel_scope:
+            try:
+                retval = await func(*args, task_status=task_status, **kwargs)
+            except BaseException as exc:
+                handle.set_exception(exc)
+            else:
+                handle.set_return_value(retval)
+
+    async def start(
+        self,
+        func: Callable[[Unpack[PosArgsT]], Coroutine[Any, Any, T]],
+        /,
+        *args: Unpack[PosArgsT],
+        kwargs: Mapping[str, Any] | None = None,
+    ) -> TaskHandle[T]:
+        handle = TaskHandle[T]()
+        handle._start_value = await self._task_group.start(
```

Comment:
If we want to be able to await the start value in the `TaskHandle`, I guess we should wrap `start()` with a `create_task()`?

Reply:
I don't understand what you mean. The start value will already be available in the `TaskHandle` once `start()` returns.

Reply:
Ah indeed :+1:

Reply:
I thought that the `TaskHandle` was returned immediately, but it's returned only when the task has started. Which means we cannot e.g. cancel the task before it has started. Not sure if that should be allowed?